src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,444 @@
+     * Whether or not this [IceTransport] has connected.
+     */
+    private val iceConnected = AtomicBoolean(false)
+    /**
+     * Whether or not this [IceTransport] has failed to connect.
+     */
+    private val iceFailed = AtomicBoolean(false)
+
+    fun hasFailed(): Boolean = iceFailed.get()
+
+    fun isConnected(): Boolean = iceConnected.get()
+
+    /**
+     * Whether or not this transport is 'running'.  If it is not
+     * running, no more data will be read from the socket or sent out.
+     */
+    private val running = AtomicBoolean(true)
+
+    private val iceAgent = Agent(Config.ufragPrefix(), logger).apply {
+        appendHarvesters(this)
+        isControlling = controlling
+        performConsentFreshness = true
+        addStateChangeListener(this@IceTransport::iceStateChanged)
+    }.also {
+        logger.addContext("local_ufrag", it.localUfrag)
+    }
+
+    // TODO: Do we still need the id here now that we have logContext?
+    private val iceStream = iceAgent.createMediaStream("stream-$id").apply {
+        addPairChangeListener(this@IceTransport::iceStreamPairChanged)
+    }
+
+    private val iceComponent = iceAgent.createComponent(
+            iceStream,
+            Transport.UDP,
+            -1,
+            -1,
+            -1,
+            Config.keepAliveStrategy(),
+            Config.useComponentSocket()
+    )
+
+    private val packetStats = PacketStats()
+
+    val icePassword: String
+        get() = iceAgent.localPassword
+
+    /**
+     * Tell this [IceTransport] to start ICE connectivity establishment.
+     */
+    fun startConnectivityEstablishment(transportPacketExtension: IceUdpTransportPacketExtension) {
+        if (!running.get()) {
+            logger.warn("Not starting connectivity establishment, transport is not running")
+            return
+        }
+        if (iceAgent.state.isEstablished) {
+            logger.cdebug { "Connection already established" }
+            return
+        }
+        logger.cdebug { "Starting ICE connectivity establishment" }
+
+        // Set the remote ufrag/password
+        iceStream.remoteUfrag = transportPacketExtension.ufrag
+        iceStream.remotePassword = transportPacketExtension.password
+
+        // If ICE is running already, we try to update the checklists with the
+        // candidates. Note that this is a best effort.
+        val iceAgentStateIsRunning = IceProcessingState.RUNNING == iceAgent.state
+
+        val remoteCandidates = transportPacketExtension.getChildExtensionsOfType(CandidatePacketExtension::class.java)
+        if (iceAgentStateIsRunning && remoteCandidates.isEmpty()) {
+            logger.cdebug { "Ignoring transport extensions with no candidates, " +
+                    "the Agent is already running." }
+            return
+        }
+
+        val remoteCandidateCount = addRemoteCandidates(remoteCandidates, iceAgentStateIsRunning)
+        if (iceAgentStateIsRunning) {
+            when (remoteCandidateCount) {

```

Comment:
@bgrozev do you think it's worth preserving the 'empty' case here?  I.e. the commented:
```
// XXX Effectively, the check above but realizing that all
// candidates were ignored:
// iceAgentStateIsRunning && candidates.isEmpty().
```
Or should I just get rid of it and turn this into a:
```
if (remoteCandidateCount != 0) {
    iceComponent.updateRemoteCandidates()
}
```
?

Reply:
I don't think we need to preserve this case